compiler/src/test/java/com/squareup/anvil/compiler/dagger/BindsMethodValidatorTest.kt

```diff
@@ -376,6 +377,36 @@ class BindsMethodValidatorTest(
     }
   }
 
+  @Test
+  fun `binding which supertype is narrower than return type fails to compile`() {
+    compile(
+      """
+        package com.squareup.test
+
+        import dagger.Module
+        import dagger.Binds
+    
+        sealed interface ItemDetail {
+          object DetailTypeA : ItemDetail
+        }
+
+        interface ItemMapper<T : ItemDetail>
+
+        class DetailTypeAItemMapper : ItemMapper<ItemDetail.DetailTypeA>
+        
+        @Module
+        interface SomeModule {
+          @Binds fun shouldBeInvalidComplexBinding(real: DetailTypeAItemMapper): ItemMapper<ItemDetail>
+        }
+      """.trimIndent(),
+    ) {
+      assertThat(exitCode).isEqualTo(ExitCode.COMPILATION_ERROR)
+      assertThat(messages).contains(
+        "@Binds methods' parameter type must be assignable to the return type",
+      )
```

Comment:
Please also include an assert here verifying the more detailed portion of the error message that calls out the specific types that didn't match up. See `a binding with an incompatible parameter type fails to compile` for an example